modules/standard/List.chpl

```diff
@@ -1318,20 +1324,152 @@ module List {
       return;
     }
 
+    /*
+      Return a copy of the element at a given index in this list.
+
+      :arg i: The index of the element to get.
+
+      .. warning::
+
+        Use of the `getValue` method with an out of bounds index (while
+        bounds checking is on) will cause the currently running program
+        to halt.
+
+      :return: A copy of an element from this list.
+    */
+    proc const getValue(i: int): eltType {
+      _enter(); defer _leave();
+
+      if boundsChecking && !_withinBounds(i) {
+        const msg = "Invalid list index: " + i:string;
+        boundsCheckHalt(msg);
+      }
+
+      return _getRef(i);
+    }
+
+    /*
+      Return a borrow of the element at a given index in this list. This
+      method can only be called when this list's element type is a class
+      type.
+
+      :arg i: The index of the element to borrow.
+      :type i: `int`
+
+      :return: A borrow of an element from this list.
+    */
+    proc const getBorrowed(i: int) where isClass(eltType) {
+      _enter(); defer _leave();
+
+      if boundsChecking && !_withinBounds(i) {
+        const msg = "Invalid list index: " + i:string;
+        boundsCheckHalt(msg);
+      }
+
+      ref slot = _getRef(i);
+
+      return slot.borrow();
+    }
+
+    /*
+      Sets the element at a given index in this list. This method returns
+      `false` if the index is out of bounds.
+
+      :arg i: The index of the element to set
+      :type i: int
+
+      :arg x: The value to set at index `i`
+
+      :return: `true` if `i` is a valid index that has been set to `x`,
+               and `false` otherwise.
+      :rtype: bool
+    */
+    proc ref set(i: int, pragma "no auto destroy" in x: eltType): bool {
+      _enter(); defer _leave();
+
+      if !_withinBounds(i) {
+        _destroy(x);
+        return false;
+      }
+
+      ref src = x;
+      ref dst = _getRef(i);
+      _move(src, dst);
+
+      return true;
+    }
+
+    /*
+      Update a value in this list in a parallel safe manner via a worker
+      object.
+
+      The worker object passed to the `update()` method must
+      define a `this()` method that takes two arguments: an integer index,
+      and a second argument of this list's `valType`. The worker object's
+      `this()` method must return some sort of value. Workers that do not
+      need to return anything may return `none`.
+
+      If the worker's `this()` method throws, the thrown error will be
+      propagated out of `update()`.
+
+      :arg i: The index to update
+      :type i: `int`
+
+      :arg worker: A class or record used to update the value at `i`
+      :return: What the worker returns
+    */
+    proc update(i: int, worker) throws {
```

Comment:
`worker` seems to me to be an odd name for this argument. `worker` might imply something about tasks to people.

Reply:
Suggestion? I'm drawing a blank.

Reply:
maybe "updater" or "operation" or "fn"

Reply:
I've gone with the term `updater`, for lack of something better.